llvm/lib/Support/Unix/Memory.inc

```diff
@@ -177,6 +181,78 @@ std::error_code Memory::protectMappedMemory(const MemoryBlock &M,
       alignAddr((const uint8_t *)M.Address + M.AllocatedSize, PageSize);
 
   bool InvalidateCache = (Flags & MF_EXEC);
+  bool SkipMprotect = false;
+
+#if defined(__linux__)
+  // Check for cases where the EXEC protection flag changes and a possible
+  // strict W^X policy cannot be bypassed via mprotect() alone, e.g. under
+  // PaX's MPROTECT or SELinux's deny_execmem.
+  //
+  // To support such systems, we need to create a fresh mapping with the
+  // target protection flags.
+  if ((M.Flags ^ Flags) & MF_EXEC && execProtChangeNeedsNewMapping()) {
+    static unsigned int flags = MFD_CLOEXEC | MFD_EXEC;
+    class FDWrapper {
```

Comment:
Weirdly, there is a wrapper like this for locking in `llvm/include/llvm/Support/FileSystem.h` (FileLocker), but none for closing.

Reply:
Certainly you should look at using the APIs in FileSystem.h instead of the raw calls.

Reply:
Hmm, `raw_fd_ostream` might be something. Albeit it still does a little bit too much (` lseek()` and `fstat()` in its constructor). But it's close, will look further. Thanks for making me look again!